File: Rust/src/donp/observe/statistics.rs
/*
	Module 'observe' provides mechanism's for observing and logging system statistics during application runtime.
*/

/*
	The Statistics module provides methods to track and log timing statistics during application execution.
*/

use log::info;

pub struct Statistics {
    start_time: std::time::Instant,
    end_time: std::time::Instant,
}


impl Statistics {
    pub fn new() -> Statistics {
        let stats = Statistics {
            start_time: std::time::Instant::now(),
            end_time: std::time::Instant::now(),
        };
        stats
    }

    pub fn start(&mut self) {
        // Start timing by resetting start and end times.
        self.reset();
    }

    pub fn end(&mut self) {
        // End timing by recording the end time.
        self.end_time = std::time::Instant::now();
    }

    pub fn get_elapsed_time_seconds(&self) -> f32 {
        // Calculate elapsed time in seconds between start and end times.
        self.end_time.duration_since(self.start_time).as_secs_f32()
    }

    pub fn log(&self) {
        // Log the elapsed time in seconds.
        info!("Statistics: Elapsed time={:?} seconds", self.get_elapsed_time_seconds());
    }

     pub fn reset(&mut self) {
        // Reset start and end times to the current instant.
        self.start_time = std::time::Instant::now();
        self.end_time = std::time::Instant::now();
    }
}

File: C++/pkg/observe/statistics.cpp
/*
	Module 'observe' provides mechanism's for observing and logging system statistics during application runtime.
*/

/*
	The Statistics module provides methods to track and log timing statistics during application execution.
*/

#include "pkg/observe/statistics.hpp"


// Constructor
Statistics::Statistics() : start_time(0), end_time(0), elapsed_time(0.0), logger() {}

// Destructor
Statistics::~Statistics() {}

void Statistics::start() {
    start_time = std::chrono::duration_cast<std::chrono::nanoseconds>(
        std::chrono::high_resolution_clock::now().time_since_epoch()
    ).count();
}

void Statistics::end() {
    end_time = std::chrono::duration_cast<std::chrono::nanoseconds>(
        std::chrono::high_resolution_clock::now().time_since_epoch()
    ).count();
}

double Statistics::get_elapsed_time_seconds() {
    return (end_time - start_time) / 1e9;
}

void Statistics::log() {
    // Log at info level with elapsed time in seconds.
    std::string message = "Statistics: Elapsed time=" + std::to_string(get_elapsed_time_seconds()) + " seconds";
    logger.log(INFO, message);
}

void Statistics::reset() {
    start_time = 0;
    end_time = 0;
    elapsed_time = 0.0;
}